File: Sources/VoxLocalCore/Transcription/WhisperOutputParser.swift
import Foundation

/// Parses whisper-cli JSON output (`-oj`) and optionally strips common
/// non-speech annotations that Whisper emits.
public enum WhisperOutputParser {
    public struct Output: Equatable, Sendable {
        public let text: String
        public let detectedLanguage: String?

        public init(text: String, detectedLanguage: String?) {
            self.text = text
            self.detectedLanguage = detectedLanguage
        }
    }

    private struct WhisperJSON: Decodable {
        struct Segment: Decodable { let text: String }
        struct ResultInfo: Decodable { let language: String? }
        let transcription: [Segment]?
        let result: ResultInfo?
    }

    public static func parse(jsonData: Data, removeArtifacts: Bool) throws -> Output {
        let decoded = try JSONDecoder().decode(WhisperJSON.self, from: jsonData)
        let joined = (decoded.transcription ?? [])
            .map(\.text)
            .joined()
        let cleaned = removeArtifacts ? stripArtifacts(from: joined) : joined
        let text = normalizeWhitespace(cleaned)
        return Output(text: text, detectedLanguage: decoded.result?.language)
    }

    // MARK: - Artifact removal

    /// Square-bracket annotations are never real dictation output.
    private static let bracketPattern = try! NSRegularExpression(pattern: #"\[[^\]\n]{0,60}\]"#)

    /// Parenthesised annotations are removed only when they match known
    /// noise markers, so real speech like "(so to speak)" survives.
    private static let knownParenMarkers: [String] = [
        "music", "applause", "laughter", "laughing", "typing", "silence",
        "inaudible", "noise", "coughing", "sighs", "sigh", "beep", "clicking",
        "музыка", "аплодисменты", "смех", "тишина", "шум", "неразборчиво",
        "вздох", "кашель", "щелчок",
    ]
    private static let parenPattern = try! NSRegularExpression(
        pattern: #"\((?:"# + knownParenMarkers.joined(separator: "|") + #")[^)\n]{0,20}\)"#,
        options: [.caseInsensitive])

    public static func stripArtifacts(from text: String) -> String {
        var result = text
        for regex in [bracketPattern, parenPattern] {
            let range = NSRange(result.startIndex..., in: result)
            result = regex.stringByReplacingMatches(in: result, range: range, withTemplate: " ")
        }
        result = result.replacingOccurrences(of: "♪", with: " ")
        return result
    }

    public static func normalizeWhitespace(_ text: String) -> String {
        let collapsed = text
            .components(separatedBy: .newlines)
            .map { $0.split(separator: " ", omittingEmptySubsequences: true).joined(separator: " ") }
            .filter { !$0.isEmpty }
            .joined(separator: "\n")
        return collapsed.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: windows/src/VoxLocal.App/Core/Transcription/WhisperOutputParser.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace VoxLocal.Core.Transcription;

/// <summary>
/// Parses whisper-cli JSON output (-oj) and optionally strips common
/// non-speech annotations that Whisper emits.
/// </summary>
public static class WhisperOutputParser
{
    public sealed record Output(string Text, string? DetectedLanguage);

    private sealed class WhisperJson
    {
        [JsonPropertyName("transcription")]
        public List<Segment>? Transcription { get; set; }

        [JsonPropertyName("result")]
        public ResultInfo? Result { get; set; }

        public sealed class Segment
        {
            [JsonPropertyName("text")]
            public string Text { get; set; } = "";
        }

        public sealed class ResultInfo
        {
            [JsonPropertyName("language")]
            public string? Language { get; set; }
        }
    }

    public static Output Parse(byte[] jsonData, bool removeArtifacts)
    {
        var decoded = JsonSerializer.Deserialize<WhisperJson>(jsonData)
            ?? throw new JsonException("empty whisper JSON");
        var joined = string.Concat((decoded.Transcription ?? new()).Select(s => s.Text));
        var cleaned = removeArtifacts ? StripArtifacts(joined) : joined;
        var text = NormalizeWhitespace(cleaned);
        return new Output(text, decoded.Result?.Language);
    }

    // ---- Artifact removal ----

    /// <summary>Square-bracket annotations are never real dictation output.</summary>
    private static readonly Regex BracketPattern = new(@"\[[^\]\n]{0,60}\]", RegexOptions.Compiled);

    /// <summary>
    /// Parenthesised annotations are removed only when they match known
    /// noise markers, so real speech like "(so to speak)" survives.
    /// </summary>
    private static readonly string[] KnownParenMarkers =
    {
        "music", "applause", "laughter", "laughing", "typing", "silence",
        "inaudible", "noise", "coughing", "sighs", "sigh", "beep", "clicking",
        "музыка", "аплодисменты", "смех", "тишина", "шум", "неразборчиво",
        "вздох", "кашель", "щелчок",
    };

    private static readonly Regex ParenPattern = new(
        @"\((?:" + string.Join("|", KnownParenMarkers) + @")[^)\n]{0,20}\)",
        RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    public static string StripArtifacts(string text)
    {
        var result = BracketPattern.Replace(text, " ");
        result = ParenPattern.Replace(result, " ");
        result = result.Replace("♪", " ");
        return result;
    }

    public static string NormalizeWhitespace(string text)
    {
        var collapsed = string.Join("\n",
            text.Split('\n', '\r')
                .Select(line => string.Join(" ", line.Split(' ', '\t').Where(w => w.Length > 0)))
                .Where(line => line.Length > 0));
        return collapsed.Trim();
    }
}
